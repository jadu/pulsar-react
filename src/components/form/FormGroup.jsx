import React from 'react';
import classnames from 'classnames';
import shortid from 'shortid';

import Fieldset from './Fieldset';
import FormLabel from './FormLabel';
import ErrorBlock from './ErrorBlock';
import HelpBlock from './HelpBlock';

export default class FormGroup extends React.Component {

  static defaultProps = {
    className: 'form__group',
    required: false
  };

  componentWillMount() {
    this.guid = shortid.generate();
  }

  render() {
    if (!this.props.children) {
      return null;
    }

    let {
      appendText,
      buttonGroup,
      changed,
      checkbox,
      children,
      choiceGroup,
      className,
      error,
      flushLabel,
      helpText,
      indented,
      inlineCheckbox,
      inlineRadioButton,
      labelText,
      prependText,
      required,
      radio,
      success,
      toggle,
      topLabel,
      warning,
      width,
      ...props} = this.props;

    // Convert variants to their required classes
    let variantClasses = classnames(className, {
      [`form__control-col--${width}`]: width,
      'form__button-group': buttonGroup,
      'form-choice': choiceGroup,
      'form__group--indent': indented,
      'form__group--toggle': toggle,
      'form__group--top': topLabel,
      'form__group--flush': flushLabel,
      'form-checkbox': checkbox,
      'form-checkbox-inline': inlineCheckbox,
      'form-radio-inline': inlineRadioButton,
      'form-radio': radio,
      'has-changed': changed,
      'has-success': success,
      'has-warning': warning,
      'has-error': error
    });

    // GUID to use if no explicit ID has been set
    let idGuid = 'id-guid-' + this.guid;
    
    // GUID to use to link the input with any help text 
    let helpGuid = 'help-guid-' + this.guid;
    
    // GUID to use to link the input with any errors 
    let errorGuid = 'error-guid-' + this.guid;

    // Build list of IDs for the input based on the presence of help/error text 
    let ariaDescribedby = classnames({
      [`${errorGuid}`]: error,
      [`${helpGuid}`]: this.props.helpText
    });

    // If help/errors are present pass the GUIDs to the child inputs
    let childrenWithGuids = React.cloneElement(children, {
      ariaDescribedby: ariaDescribedby ? ariaDescribedby : null,
      idGuid: idGuid
    });

    // Build list of classNames to be used on the controls block
    let controlsClassName = classnames('controls', {
      'btn__group': buttonGroup
    });

    // Form controls
    let inputBlock = (
      <>{childrenWithGuids}</>
    );

    // Wrap with input group if appended/prepended text is present
    if (appendText || prependText) {
      inputBlock = (
        <div className="input-group">
          {prependText && 
            <span className="input-group-addon">{prependText}</span>
          }
          {childrenWithGuids}
          {appendText && 
            <span className="input-group-addon">{appendText}</span>
          }
        </div>
      );
    }

    // Wrap input block alongside errors and help
    let controlsBlock = (
      <div className={controlsClassName}>
        {inputBlock}
        <ErrorBlock errorGuid={errorGuid}>
          {error}
        </ErrorBlock> 
        <HelpBlock helpGuid={helpGuid}>
          {helpText}
        </HelpBlock>
      </div>
    );

    // Standard group markup strategy for most components
    let groupBlock = (
      <div className={variantClasses}>
        <FormLabel 
          required={required} 
          idGuid={idGuid}
        >
          {labelText}
        </FormLabel>
        {controlsBlock}
      </div>
    );

    // Switch markup strategies for certain non-standard components
    if (choiceGroup) {
      groupBlock = (
        <Fieldset 
          className={variantClasses}
          legendText={labelText}
          legendClassName="control__label"
        >
          {controlsBlock}
        </Fieldset>
      );
    }
    else if (toggle) {
      groupBlock = (
        <div className={variantClasses}>
          <FormLabel 
            className="toggle-switch-wrapper-label" 
            required={required} 
            idGuid={idGuid}
          >
            <span className="control__label">
              {labelText}
            </span>
            {controlsBlock}
          </FormLabel>
        </div>
      );
    } 
    else if (inlineCheckbox || inlineRadioButton) {
      groupBlock = (
        <div className={variantClasses}>
          {controlsBlock}
          {labelText}
        </div>
      );
    }

    return (
      <>{groupBlock}</>
    );
  }
}