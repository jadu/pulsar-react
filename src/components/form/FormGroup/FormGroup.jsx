import React from 'react';
import classnames from 'classnames';
import shortid from 'shortid';

import Fieldset from '../Fieldset/Fieldset';
import FormLabel from '../FormLabel/FormLabel';
import ErrorBlock from '../ErrorBlock/ErrorBlock';
import HelpBlock from '../HelpBlock/HelpBlock';

export default class FormGroup extends React.Component {

  static defaultProps = {
    className: 'form__group',
    required: false
  };

  render() {
    if (!this.props.children) {
      return null;
    }

    let {
      appendText,
      buttonGroup,
      changed,
      checkbox,
      children,
      choiceGroup,
      className,
      controls,
      error,
      flushLabel,
      helpText,
      hideLabel,
      indented,
      inline,
      inlineCheckbox,
      inlineRadioButton,
      labelText,
      prependText,
      required,
      radioButton,
      rangeInput,
      success,
      toggle,
      topLabel,
      warning,
      width,
      ...props} = this.props;

    // Convert variants to their required classes
    let variantClasses = classnames(className, {
      [`form__control-col--${width}`]: width,
      'form__button-group': buttonGroup,
      'form-choice': choiceGroup,
      'form__group--compound': inline,
      'form__group--indent': indented,
      'form__group--toggle': toggle,
      'form__group--top': topLabel,
      'form__group--flush': flushLabel,
      'form-checkbox': checkbox,
      'form-checkbox-inline': inlineCheckbox,
      'form-radio-inline': inlineRadioButton,
      'form-radio': radioButton,
      'form-range': rangeInput,
      'has-changed': changed,
      'has-success': success,
      'has-warning': warning,
      'has-error': error
    });

    let guid = this.props.guid || shortid.generate();

    // GUID to use if no explicit ID has been set
    let idGuid = 'id-guid-' + guid;
    
    // GUID to use to link the input with any help text 
    let helpGuid = 'help-guid-' + guid;
    
    // GUID to use to link the input with any errors 
    let errorGuid = 'error-guid-' + guid;

    // Build list of IDs for the input based on the presence of help/error text 
    let ariaDescribedby = classnames({
      [`${errorGuid}`]: error,
      [`${helpGuid}`]: this.props.helpText
    });

    // If help/errors are present pass the GUIDs to the child inputs
    let childrenWithGuids = React.Children.map(
      children,
      (child, i) => {
        if (['div', 'p'].indexOf(child.type) !== -1) {
          return child;
        }

        return React.cloneElement(child, {
          ariaDescribedby: ariaDescribedby ? ariaDescribedby : null,
          id: idGuid
        });
      }
    );

    // Build list of classNames to be used on the controls block
    let controlsClassName = classnames('controls', {
      'btn__group': buttonGroup
    });

    // Form controls
    let inputBlock = (
      <React.Fragment>{childrenWithGuids}</React.Fragment>
    );

    // Wrap with input group if appended/prepended text is present
    if (appendText || prependText) {
      inputBlock = (
        <div className="input-group">
          {prependText && 
            <span className="input-group-addon">{prependText}</span>
          }
          {childrenWithGuids}
          {appendText && 
            <span className="input-group-addon">{appendText}</span>
          }
        </div>
      );
    }

    // Wrap input block alongside errors and help
    let controlsBlock = (
      <div className={controlsClassName}>
        {inputBlock}
        <ErrorBlock errorGuid={errorGuid}>
          {error}
        </ErrorBlock> 
        <HelpBlock helpGuid={helpGuid}>
          {helpText}
        </HelpBlock>
      </div>
    );

    // Standard group markup strategy for most components
    let groupBlock = (
      <div className={variantClasses}>
        {labelText &&
        <FormLabel 
          required={required} 
          idGuid={idGuid}
          tag={controls ? 'span' : 'label'}
          hideLabel={hideLabel}
        >
          {labelText}
        </FormLabel>
        }

        {controlsBlock}
      </div>
    );

    // Switch markup strategies for certain non-standard components
    if (choiceGroup) {
      groupBlock = (
        <Fieldset 
          className={variantClasses}
          legendText={labelText}
          legendClassName="control__label"
        >
          {controlsBlock}
        </Fieldset>
      );
    }
    else if (toggle) {
      groupBlock = (
        <div className={variantClasses}>
          <FormLabel 
            className="toggle-switch-wrapper-label" 
            required={required} 
            idGuid={idGuid}
          >
            <span className="control__label">
              {labelText}
            </span>
            {controlsBlock}
          </FormLabel>
        </div>
      );
    } 
    else if (inlineCheckbox || inlineRadioButton) {
      groupBlock = (
        <div className={variantClasses}>
          {controlsBlock}
          {labelText}
        </div>
      );
    }

    return (
      <React.Fragment>{groupBlock}</React.Fragment>
    );
  }
}