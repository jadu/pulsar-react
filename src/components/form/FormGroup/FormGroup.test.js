'use strict';

import React from 'react';
import { shallow, render } from 'enzyme';

import FormGroup from './FormGroup';

it('should return null if no children', () => {
  const tree = render(
    <FormGroup></FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('should render the basic form group', () => {
  const tree = render(
    <FormGroup>
      <>foo</>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('should add the button group classes if using buttonGroup variant', () => {
  const tree = shallow(
    <FormGroup buttonGroup>
      <>foo</>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('should use fieldset strategy if using choiceGroup variant', () => {
  const tree = shallow(
    <FormGroup choiceGroup>
      <>foo</>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('should add the indented class if using indented variant', () => {
  const tree = shallow(
    <FormGroup indented>
      <>foo</>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('should use toggle strategy if using toggle variant', () => {
  const tree = shallow(
    <FormGroup toggle>
      <>foo</>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('should add the top label class if using topLabel variant', () => {
  const tree = shallow(
    <FormGroup topLabel>
      <>foo</>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('should add the flush label class if using flushLabel variant', () => {
  const tree = shallow(
    <FormGroup flushLabel>
      <>foo</>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('should add the checkbox class if using checkbox variant', () => {
  const tree = shallow(
    <FormGroup checkbox>
      <>foo</>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('should use inline strategy if using inlineCheckbox variant', () => {
  const tree = shallow(
    <FormGroup inlineCheckbox>
      <>foo</>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('should use inline strategy if using inlineRadioButton variant', () => {
  const tree = shallow(
    <FormGroup inlineRadioButton>
      <>foo</>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('should add the radio class if using radio variant', () => {
  const tree = shallow(
    <FormGroup radio>
      <>foo</>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('should add the success class if using success variant', () => {
  const tree = shallow(
    <FormGroup success>
      <>foo</>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('should add the changed class if using changed variant', () => {
  const tree = shallow(
    <FormGroup changed>
      <>foo</>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('should add the warning class if using warning variant', () => {
  const tree = shallow(
    <FormGroup warning>
      <>foo</>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('should add the error class if using error variant', () => {
  const tree = shallow(
    <FormGroup error="error message">
      <>foo</>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('should prepend text', () => {
  const tree = shallow(
    <FormGroup prependText="prepended">
      <>foo</>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('should append text', () => {
  const tree = shallow(
    <FormGroup appendText="appended">
      <>foo</>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});

it('should append and prepend text', () => {
  const tree = shallow(
    <FormGroup prependText="prepended" appendText="appended">
      <>foo</>
    </FormGroup>
  );
  expect(tree).toMatchSnapshot();
});