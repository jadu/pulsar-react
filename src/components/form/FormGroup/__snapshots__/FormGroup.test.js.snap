// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add the button group classes if using buttonGroup variant 1`] = `
<Fragment>
  <div
    className="form__group form__button-group"
  >
    <FormLabel
      className="control__label"
      idGuid="id-guid-1"
      required={false}
    />
    <div
      className="controls btn__group"
    >
      foo
      <ErrorBlock
        errorGuid="error-guid-1"
      />
      <HelpBlock
        helpGuid="help-guid-1"
      />
    </div>
  </div>
</Fragment>
`;

exports[`should add the changed class if using changed variant 1`] = `
<Fragment>
  <div
    className="form__group has-changed"
  >
    <FormLabel
      className="control__label"
      idGuid="id-guid-1"
      required={false}
    />
    <div
      className="controls"
    >
      foo
      <ErrorBlock
        errorGuid="error-guid-1"
      />
      <HelpBlock
        helpGuid="help-guid-1"
      />
    </div>
  </div>
</Fragment>
`;

exports[`should add the checkbox class if using checkbox variant 1`] = `
<Fragment>
  <div
    className="form__group form-checkbox"
  >
    <FormLabel
      className="control__label"
      idGuid="id-guid-1"
      required={false}
    />
    <div
      className="controls"
    >
      foo
      <ErrorBlock
        errorGuid="error-guid-1"
      />
      <HelpBlock
        helpGuid="help-guid-1"
      />
    </div>
  </div>
</Fragment>
`;

exports[`should add the error class if using error variant 1`] = `
<Fragment>
  <div
    className="form__group has-error"
  >
    <FormLabel
      className="control__label"
      idGuid="id-guid-1"
      required={false}
    />
    <div
      className="controls"
    >
      foo
      <ErrorBlock
        errorGuid="error-guid-1"
      >
        error message
      </ErrorBlock>
      <HelpBlock
        helpGuid="help-guid-1"
      />
    </div>
  </div>
</Fragment>
`;

exports[`should add the flush label class if using flushLabel variant 1`] = `
<Fragment>
  <div
    className="form__group form__group--flush"
  >
    <FormLabel
      className="control__label"
      idGuid="id-guid-1"
      required={false}
    />
    <div
      className="controls"
    >
      foo
      <ErrorBlock
        errorGuid="error-guid-1"
      />
      <HelpBlock
        helpGuid="help-guid-1"
      />
    </div>
  </div>
</Fragment>
`;

exports[`should add the indented class if using indented variant 1`] = `
<Fragment>
  <div
    className="form__group form__group--indent"
  >
    <FormLabel
      className="control__label"
      idGuid="id-guid-1"
      required={false}
    />
    <div
      className="controls"
    >
      foo
      <ErrorBlock
        errorGuid="error-guid-1"
      />
      <HelpBlock
        helpGuid="help-guid-1"
      />
    </div>
  </div>
</Fragment>
`;

exports[`should add the radio class if using radio variant 1`] = `
<Fragment>
  <div
    className="form__group form-radio"
  >
    <FormLabel
      className="control__label"
      idGuid="id-guid-1"
      required={false}
    />
    <div
      className="controls"
    >
      foo
      <ErrorBlock
        errorGuid="error-guid-1"
      />
      <HelpBlock
        helpGuid="help-guid-1"
      />
    </div>
  </div>
</Fragment>
`;

exports[`should add the success class if using success variant 1`] = `
<Fragment>
  <div
    className="form__group has-success"
  >
    <FormLabel
      className="control__label"
      idGuid="id-guid-1"
      required={false}
    />
    <div
      className="controls"
    >
      foo
      <ErrorBlock
        errorGuid="error-guid-1"
      />
      <HelpBlock
        helpGuid="help-guid-1"
      />
    </div>
  </div>
</Fragment>
`;

exports[`should add the top label class if using topLabel variant 1`] = `
<Fragment>
  <div
    className="form__group form__group--top"
  >
    <FormLabel
      className="control__label"
      idGuid="id-guid-1"
      required={false}
    />
    <div
      className="controls"
    >
      foo
      <ErrorBlock
        errorGuid="error-guid-1"
      />
      <HelpBlock
        helpGuid="help-guid-1"
      />
    </div>
  </div>
</Fragment>
`;

exports[`should add the warning class if using warning variant 1`] = `
<Fragment>
  <div
    className="form__group has-warning"
  >
    <FormLabel
      className="control__label"
      idGuid="id-guid-1"
      required={false}
    />
    <div
      className="controls"
    >
      foo
      <ErrorBlock
        errorGuid="error-guid-1"
      />
      <HelpBlock
        helpGuid="help-guid-1"
      />
    </div>
  </div>
</Fragment>
`;

exports[`should append and prepend text 1`] = `
<Fragment>
  <div
    className="form__group"
  >
    <FormLabel
      className="control__label"
      idGuid="id-guid-1"
      required={false}
    />
    <div
      className="controls"
    >
      <div
        className="input-group"
      >
        <span
          className="input-group-addon"
        >
          prepended
        </span>
        foo
        <span
          className="input-group-addon"
        >
          appended
        </span>
      </div>
      <ErrorBlock
        errorGuid="error-guid-1"
      />
      <HelpBlock
        helpGuid="help-guid-1"
      />
    </div>
  </div>
</Fragment>
`;

exports[`should append text 1`] = `
<Fragment>
  <div
    className="form__group"
  >
    <FormLabel
      className="control__label"
      idGuid="id-guid-1"
      required={false}
    />
    <div
      className="controls"
    >
      <div
        className="input-group"
      >
        foo
        <span
          className="input-group-addon"
        >
          appended
        </span>
      </div>
      <ErrorBlock
        errorGuid="error-guid-1"
      />
      <HelpBlock
        helpGuid="help-guid-1"
      />
    </div>
  </div>
</Fragment>
`;

exports[`should prepend text 1`] = `
<Fragment>
  <div
    className="form__group"
  >
    <FormLabel
      className="control__label"
      idGuid="id-guid-1"
      required={false}
    />
    <div
      className="controls"
    >
      <div
        className="input-group"
      >
        <span
          className="input-group-addon"
        >
          prepended
        </span>
        foo
      </div>
      <ErrorBlock
        errorGuid="error-guid-1"
      />
      <HelpBlock
        helpGuid="help-guid-1"
      />
    </div>
  </div>
</Fragment>
`;

exports[`should render the basic form group 1`] = `
<div
  class="form__group"
>
  <span
    class="control__label"
  />
  <div
    class="controls"
  >
    foo
  </div>
</div>
`;

exports[`should return null if no children 1`] = `null`;

exports[`should use fieldset strategy if using choiceGroup variant 1`] = `
<Fragment>
  <Fieldset
    className="form__group form-choice"
    legendClassName="control__label"
  >
    <div
      className="controls"
    >
      foo
      <ErrorBlock
        errorGuid="error-guid-1"
      />
      <HelpBlock
        helpGuid="help-guid-1"
      />
    </div>
  </Fieldset>
</Fragment>
`;

exports[`should use inline strategy if using inlineCheckbox variant 1`] = `
<Fragment>
  <div
    className="form__group form-checkbox-inline"
  >
    <div
      className="controls"
    >
      foo
      <ErrorBlock
        errorGuid="error-guid-1"
      />
      <HelpBlock
        helpGuid="help-guid-1"
      />
    </div>
  </div>
</Fragment>
`;

exports[`should use inline strategy if using inlineRadioButton variant 1`] = `
<Fragment>
  <div
    className="form__group form-radio-inline"
  >
    <div
      className="controls"
    >
      foo
      <ErrorBlock
        errorGuid="error-guid-1"
      />
      <HelpBlock
        helpGuid="help-guid-1"
      />
    </div>
  </div>
</Fragment>
`;

exports[`should use toggle strategy if using toggle variant 1`] = `
<Fragment>
  <div
    className="form__group form__group--toggle"
  >
    <FormLabel
      className="toggle-switch-wrapper-label"
      idGuid="id-guid-1"
      required={false}
    >
      <span
        className="control__label"
      />
      <div
        className="controls"
      >
        foo
        <ErrorBlock
          errorGuid="error-guid-1"
        />
        <HelpBlock
          helpGuid="help-guid-1"
        />
      </div>
    </FormLabel>
  </div>
</Fragment>
`;
